### ISSUE TITLE
Class properties type inferred as never without this prefix

### ISSUE THREAD

**Issue by OmarTawfik** [Microsoft/TypeScript#30277]
**TypeScript Version:** 
* latest (tested on the [playground](https://www.typescriptlang.org/play/#src=export%20class%20MyType%20%7B%0D%0A%20%20public%20constructor(%0D%0A%20%20%20%20private%20readonly%20name%3A%20string%2C%0D%0A%20%20)%20%7B%0D%0A%20%20%7D%0D%0A%20%20public%20compose()%3A%20string%5B%5D%20%7B%0D%0A%20%20%20%20return%20%5B%0D%0A%20%20%20%20%20%20notHere%2C%20%20%2F%2F%20ERROR%3A%20Cannot%20find%20name%20'notHere'.%0D%0A%20%20%20%20%20%20name%2C%20%20%20%20%20%2F%2F%20type%20is%20'never'%2C%20should%20be%20an%20error%20instead%20(same%20as%20above).%0D%0A%20%20%20%20%20%20this.name%20%2F%2F%20type%20is%20'string'%20as%20expected.%0D%0A%20%20%20%20%5D%3B%0D%0A%20%20%7D%0D%0A%7D))
* `3.3.3` (tested in vscode)

**Search Terms:**
* Using a class property without this keyword.
* Class properties type inferred as `never`.

**Code**

```ts
export class MyType {
  public constructor(
    private readonly name: string,
  ) {
  }
  public compose(): string[] {
    return [
      notHere,  // ERROR: Cannot find name 'notHere'.
      name,     // type is 'never', no errors are reported.
      this.name // type is 'string' as expected.
    ];
  }
}
```

**Expected behavior:**

1. Using `name` without prefixing it with `this.` should report the following error: `Cannot find name 'name'. Did you mean the instance member 'this.name'?`
2. Type inference should still succeed, and infer `string` for this expression.

**Actual behavior:**

No error on using `name`. Just type is silently inferred as `never`.

**Playground Link:** [click here](https://www.typescriptlang.org/play/#src=export%20class%20MyType%20%7B%0D%0A%20%20public%20constructor(%0D%0A%20%20%20%20private%20readonly%20name%3A%20string%2C%0D%0A%20%20)%20%7B%0D%0A%20%20%7D%0D%0A%20%20public%20compose()%3A%20string%5B%5D%20%7B%0D%0A%20%20%20%20return%20%5B%0D%0A%20%20%20%20%20%20notHere%2C%20%20%2F%2F%20ERROR%3A%20Cannot%20find%20name%20'notHere'.%0D%0A%20%20%20%20%20%20name%2C%20%20%20%20%20%2F%2F%20type%20is%20'never'%2C%20should%20be%20an%20error%20instead%20(same%20as%20above).%0D%0A%20%20%20%20%20%20this.name%20%2F%2F%20type%20is%20'string'%20as%20expected.%0D%0A%20%20%20%20%5D%3B%0D%0A%20%20%7D%0D%0A%7D)

**Comment by OmarTawfik** [comment]
In fact, this error is reported even without the property defined:

```ts
export class MyType {
  public compose(): string[] {
    return [
      notHere,  // ERROR: Cannot find name 'notHere'.
      name,     // type is 'never', no errors are reported.
    ];
  }
}
```

**Comment by RyanCavanaugh** [comment]
There is a global variable called `name` in browser environments. See #18433

**Comment by OmarTawfik** [comment]
@RyanCavanaugh thanks for pointing out. Will close this one as it is not adding more information to the original one.
